### PR TITLE
BB-1 Implemented method "persist()

### PR DESCRIPTION
**Pull request by JJJazl**
* Implemented method "persist()"
* Added a few util methods
* Added SqlQueryUtil class

### CODE REVIEW

File: src/main/java/org/svydovets/dao/GenericJdbcDAO.java [hunk 1]
```diff
@@ -23,6 +20,48 @@ public GenericJdbcDAO(DataSource dataSource) {
...
+            return save(entity, connection);
+        } catch (SQLException e) {
+            throw new DaoOperationException(String.format(
+                    "Error saving entity to the DB: %s", entity.getClass().getName()),
+                    e
+            );
+        }
+    }
+
+    private Object save(Object entity, Connection connection) throws SQLException {
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/dao/GenericJdbcDAO.java:37]: Why is it necessary to move this functionality into a separate method?

File: src/main/java/org/svydovets/dao/GenericJdbcDAO.java [hunk 2]
```diff
@@ -23,6 +20,48 @@ public GenericJdbcDAO(DataSource dataSource) {
...
+            log.info(String.format("Insert: %s", insertQuery));
+        }
+        try {
+            PreparedStatement insertStatement = connection.prepareStatement(insertQuery, Statement.RETURN_GENERATED_KEYS);
+            Field[] entityFields = ReflectionUtils.getInsertableFieldsForIdentityGenerationType(entity.getClass());
+            for (int i = 0; i < entityFields.length; i++) {
+                insertStatement.setObject(i + 1, ReflectionUtils.getFieldValue(entity, entityFields[i]));
+            }
+            return insertStatement;
+        } catch (SQLException e) {
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/dao/GenericJdbcDAO.java]: do not use variable names less than 3 characters long

File: src/main/java/org/svydovets/dao/GenericJdbcDAO.java [hunk 3]
```diff
@@ -23,6 +20,48 @@ public GenericJdbcDAO(DataSource dataSource) {
...
+            throw new DaoOperationException(String.format(
+                    "Error preparing insert statement for entity: %s", entity.getClass().getName()),
+                    e
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/dao/GenericJdbcDAO.java]: see above

File: src/main/java/org/svydovets/dao/GenericJdbcDAO.java [hunk 4]
```diff
@@ -45,35 +84,32 @@ public void update(Map.Entry<EntityKey<?>, Object> keyEntityEntry) {
...
+        PreparedStatement updateByIdStatement = prepareUpdateStatement(connection, entityKey, entity);
         var updatedRowsCount = updateByIdStatement.executeUpdate();
         if (updatedRowsCount == 0) {
-            throw new DaoOperationException(String.format("Update has not been perform for entity: %s", entry.getKey()));
+            throw new DaoOperationException(String.format("Update has not been perform for entity: %s", entityKey));
         }
     }
 
-    private PreparedStatement prepareUpdateStatement(Connection connection, Map.Entry<EntityKey<?>, Object> entry) {
+    private PreparedStatement prepareUpdateStatement(Connection connection, EntityKey<?> entityKey, Object entity) {
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/dao/GenericJdbcDAO.java:133]: Why is this variable needed? -> EntityKey<?> entityKey

entity.getClass() the same ?

File: src/main/java/org/svydovets/query/SqlQueryBuilder.java [hunk 5]
```diff
@@ -14,10 +13,22 @@
 public class SqlQueryBuilder {
     private static final String SELECT_BY_ID_SQL = "select * from %s where %s = ?";
 
+    private static final String INSERT_SQL = "insert into %s(%s) values(%s)";
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/query/SqlQueryBuilder.java]: add space between  %s(%s) -> %s (%s)

File: src/main/java/org/svydovets/util/ReflectionUtils.java [hunk 6]
```diff
@@ -38,4 +39,34 @@ public static Field getIdField(Class<?> clazz) {
...
+
+    public static Field[] getInsertableFieldsForIdentityGenerationType(Class<?> clazz) {
+        return getUpdatableFields(clazz);
+    }
+
+    public static void setFieldValue(Object entity, Field entityField, Object value) {
+        try {
+            entityField.setAccessible(true);
+            entityField.set(entity, value);
+        } catch (IllegalAccessException e) {
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/util/ReflectionUtils.java:51]: do not use variable names less than 3 characters long

File: src/main/java/org/svydovets/util/ReflectionUtils.java [hunk 7]
```diff
@@ -38,4 +39,34 @@ public static Field getIdField(Class<?> clazz) {
...
+            throw new BibernateException(String.format(
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/util/ReflectionUtils.java:52]: best practice

catch (IllegalAccessException exception) {
    throw new BibernateException(MessageError, exception);
}

File: src/main/java/org/svydovets/util/ReflectionUtils.java [hunk 8]
```diff
@@ -38,4 +39,34 @@ public static Field getIdField(Class<?> clazz) {
...
+            );
+        }
+    }
+
+    public static Object getFieldValue(Object entity, Field entityField) {
+        try {
+            entityField.setAccessible(true);
+            return entityField.get(entity);
+        } catch (IllegalAccessException e) {
+            throw new BibernateException(String.format(
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/util/ReflectionUtils.java:65]: best practice

catch (IllegalAccessException exception) {
throw new BibernateException(MessageError, exception);
}

File: src/main/java/org/svydovets/util/ReflectionUtils.java [hunk 9]
```diff
@@ -38,4 +39,34 @@ public static Field getIdField(Class<?> clazz) {
...
+                    entity.getClass().getName())
+            );
+        }
+    }
+
+    public static Object getFieldValue(Object entity, Field entityField) {
+        try {
+            entityField.setAccessible(true);
+            return entityField.get(entity);
+        } catch (IllegalAccessException e) {
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/util/ReflectionUtils.java:64]: do not use variable names less than 3 characters long

File: src/main/java/org/svydovets/util/SqlQueryUtil.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+public class SqlQueryUtil {
+
+    public static String resolveColumnNamesForInsert(Class<?> clazz) {
+        Field[] insertableFields = ReflectionUtils.getInsertableFieldsForIdentityGenerationType(clazz);
+        return Arrays.stream(insertableFields)
+                .map(ParameterNameResolver::resolveColumnName)
+                .collect(Collectors.joining(", "));
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/util/SqlQueryUtil.java:15]: ", " to constant

File: src/main/java/org/svydovets/util/SqlQueryUtil.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public static String resolveColumnValuesForInsert(Class<?> clazz) {
+        Field[] insertableFields = ReflectionUtils.getInsertableFieldsForIdentityGenerationType(clazz);
+
+        return Arrays.stream(insertableFields)
+                .map(field -> "?")
+                .collect(Collectors.joining(", "));
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/util/SqlQueryUtil.java:23]: see above

File: src/main/java/org/svydovets/util/SqlQueryUtil.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public static String resolveUpdatableColumnsWithValues(Class<?> clazz) {
+        StringBuilder updateBuilder = new StringBuilder();
+        Field[] entityFields = ReflectionUtils.getUpdatableFields(clazz);
+        for (int i = 0; i < entityFields.length; i++) {
+            String columnName = ParameterNameResolver.resolveColumnName(entityFields[i]);
+            updateBuilder.append(columnName).append(" = ").append("?");
+            if (i < entityFields.length - 1) {
+                updateBuilder.append(", ");
```
Comment (AlexandrNavozenko) [inline src/main/java/org/svydovets/util/SqlQueryUtil.java:33]: see above

File: src/main/java/org/svydovets/exception/BibernateException.java [hunk 13]
```diff
@@ -0,0 +1,8 @@
+package org.svydovets.exception;
+
+public class BibernateException extends RuntimeException {
```
Comment (Pan14ek) [inline src/main/java/org/svydovets/exception/BibernateException.java:3]: Let's add Javadoc for this exception

**Pull request merged**